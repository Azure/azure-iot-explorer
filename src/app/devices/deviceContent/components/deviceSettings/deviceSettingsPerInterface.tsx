/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import { Overlay } from 'office-ui-fabric-react/lib/Overlay';
import { TooltipHost } from 'office-ui-fabric-react/lib/Tooltip';
import DeviceSettingPerInterfacePerSetting, { TwinWithSchema } from './deviceSettingsPerInterfacePerSetting';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { PatchDigitalTwinInterfacePropertiesActionParameters } from '../../actions';
import { INFO, InterfaceDetailCard } from '../../../../constants/iconNames';
import '../../../../css/_devicePnpDetailList.scss';

export interface DeviceSettingDataProps {
    deviceId: string;
    interfaceId: string;
    interfaceName: string;
    twinWithSchema: TwinWithSchema[];
}

export interface DeviceSettingDispatchProps {
    patchDigitalTwinInterfaceProperties: (parameters: PatchDigitalTwinInterfacePropertiesActionParameters) => void;
}

export interface DeviceSettingState {
    collapseMap: Map<number, boolean>;
    allCollapsed: boolean;
    showOverlay: boolean;
}

export default class DeviceSettingsPerInterface
    extends React.Component<DeviceSettingDataProps & DeviceSettingDispatchProps, DeviceSettingState> {
    constructor(props: DeviceSettingDataProps & DeviceSettingDispatchProps) {
        super(props);

        const { twinWithSchema } = this.props;
        const collapseMap = new Map();
        for (let index = 0; index < twinWithSchema.length; index ++) {
            collapseMap.set(index, false);
        }
        this.state = {
            allCollapsed: false,
            collapseMap,
            showOverlay: false
        };
    }

    public render(): JSX.Element {
        const { twinWithSchema} = this.props;

        const settings = twinWithSchema && twinWithSchema.map((tuple, index) => (
            <DeviceSettingPerInterfacePerSetting
                key={index}
                {...tuple}
                {...this.props}
                collapsed={this.state.collapseMap.get(index)}
                handleCollapseToggle={this.handleCollapseToggle(index)}
                handleOverlayToggle={this.handleOverlayToggle}
            />
        ));

        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <div className="pnp-detail-list scrollable-lg ms-Grid">
                        <div className="list-header ms-Grid-row">
                            <span className="ms-Grid-col ms-sm3">{context.t(ResourceKeys.deviceSettings.columns.name)}</span>
                            <span className="ms-Grid-col ms-sm2">{context.t(ResourceKeys.deviceSettings.columns.schema)}</span>
                            <span className="ms-Grid-col ms-sm2">{context.t(ResourceKeys.deviceSettings.columns.unit)}</span>
                            <span className="ms-Grid-col ms-sm4 reported-value">
                                {context.t(ResourceKeys.deviceSettings.columns.reportedValue)}
                                <TooltipHost
                                    content={context.t(ResourceKeys.deviceSettings.columns.reportedValueTooltip)}
                                    calloutProps={{ gapSpace: 0 }}
                                    styles={{ root: { display: 'inline-flex'} }}
                                >
                                    <IconButton
                                        iconProps={{ iconName: INFO }}
                                        ariaLabel={INFO}
                                    />
                                </TooltipHost>
                            </span>
                            {this.renderCollapseAllButton(context)}
                        </div>
                        <section role={twinWithSchema && twinWithSchema.length === 0 ? 'main' : 'list'} className="list-content">
                            {settings}
                        </section>
                        {this.state.showOverlay && <Overlay/>}
                    </div>
                )}
            </LocalizationContextConsumer>
        );
    }

    private readonly renderCollapseAllButton = (context: LocalizationContextInterface) => {
        return (
            <div className="ms-Grid-col ms-sm1 collapse-button">
                <IconButton
                    iconProps={{iconName: this.state.allCollapsed ? InterfaceDetailCard.OPEN : InterfaceDetailCard.CLOSE}}
                    ariaLabel={this.state.allCollapsed ?
                        context.t(ResourceKeys.deviceSettings.command.expandAll) :
                        context.t(ResourceKeys.deviceSettings.command.collapseAll)}
                    onClick={this.onToggleCollapseAll}
                    title={context.t(this.state.allCollapsed ? ResourceKeys.deviceSettings.command.expandAll : ResourceKeys.deviceSettings.command.collapseAll)}
                />
            </div>
        );
    }

    private readonly onToggleCollapseAll = () => {
        const allCollapsed = this.state.allCollapsed;
        const collapseMap = new Map();
        for (let index = 0; index < this.state.collapseMap.size; index ++) {
            collapseMap.set(index, !allCollapsed);
        }
        this.setState({allCollapsed: !allCollapsed, collapseMap});
    }

    private readonly handleCollapseToggle = (index: number) => () => {
        const collapseMap = this.state.collapseMap;
        collapseMap.set(index, !collapseMap.get(index));
        this.setState({collapseMap});
    }

    private readonly handleOverlayToggle = () => {
        this.setState({showOverlay: !this.state.showOverlay});
    }
}
