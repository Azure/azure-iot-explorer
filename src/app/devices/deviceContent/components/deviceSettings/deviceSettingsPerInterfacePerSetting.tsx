/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Label } from 'office-ui-fabric-react/lib/Label';
import { Stack } from 'office-ui-fabric-react/lib/Stack';
import { ActionButton, IconButton } from 'office-ui-fabric-react/lib/Button';
import { ParsedJsonSchema } from '../../../../api/models/interfaceJsonParserOutput';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { InterfaceDetailCard } from '../../../../constants/iconNames';
import { PropertyContent } from '../../../../api/models/modelDefinition';
import ComplexReportedFormPanel from '../shared/complexReportedFormPanel';
import DataForm from '../shared/dataForm';
import { RenderSimplyTypeValue } from '../shared/simpleReportedSection';
import { RenderDesiredState } from '../shared/desiredStateStatus';
import { PatchDigitalTwinInterfacePropertiesActionParameters } from '../../actions';
import { generateInterfacePropertiesPayload } from '../../sagas/digitalTwinInterfacePropertySaga';
import { Reported } from '../../../../api/models/digitalTwinModels';
import ErrorBoundary from '../../../errorBoundary';
import { getLocalizedData } from '../../../../api/dataTransforms/modelDefinitionTransform';
import '../../../../css/_deviceSettings.scss';

export interface DeviceSettingDataProps extends TwinWithSchema {
    collapsed: boolean;
    deviceId: string;
    interfaceId: string;
    interfaceName: string;
}

export interface DeviceSettingDispatchProps {
    handleCollapseToggle: () => void;
    handleOverlayToggle: () => void;
    patchDigitalTwinInterfaceProperties: (parameters: PatchDigitalTwinInterfacePropertiesActionParameters) => void;
}

export interface TwinWithSchema {
    desiredTwin: any; // tslint:disable-line: no-any
    reportedTwin: Reported;

    settingModelDefinition: PropertyContent;
    settingSchema: ParsedJsonSchema;
}

interface DeviceSettingState {
    showReportedValuePanel: boolean;
}

export default class DeviceSettingsPerInterfacePerSetting
    extends React.Component<DeviceSettingDataProps & DeviceSettingDispatchProps, DeviceSettingState> {
    constructor(props: DeviceSettingDataProps & DeviceSettingDispatchProps) {
        super(props);

        this.state = {
            showReportedValuePanel: false
        };
    }

    public render(): JSX.Element {
        return (
            <article className="list-item" role="listitem">
                <LocalizationContextConsumer>
                    {(context: LocalizationContextInterface) => (
                        <ErrorBoundary error={context.t(ResourceKeys.errorBoundary.text)}>
                            <div className="ms-Grid-row">
                                {this.createCollapsedSummary(context)}
                                {this.createUncollapsedCard()}
                                {this.createReportedValuePanel()}
                            </div>
                        </ErrorBoundary>
                    )}
                </LocalizationContextConsumer>
            </article>
        );
    }

    private readonly isSchemaSimpleType = () => {
        return this.props.settingSchema && (
            typeof this.props.settingModelDefinition.schema === 'string' ||
            this.props.settingModelDefinition.schema['@type'].toLowerCase() === 'enum');
    }

    private readonly createCollapsedSummary = (context: LocalizationContextInterface) => {
        return (
            <header className={this.props.collapsed ? 'item-summary' : 'item-summary item-summary-uncollapsed'} onClick={this.props.handleCollapseToggle}>
                {this.renderPropertyName(context)}
                {this.renderPropertySchema(context)}
                {this.renderPropertyUnit(context)}
                {this.renderPropertyReportedValue(context)}
                {this.renderCollapseButton(context)}
            </header>
        );
    }

    private readonly renderPropertyName = (context: LocalizationContextInterface) => {
        const ariaLabel = context.t(ResourceKeys.deviceSettings.columns.name);
        let displayName = getLocalizedData(this.props.settingModelDefinition.displayName);
        displayName = displayName ? displayName : '--';
        let description = getLocalizedData(this.props.settingModelDefinition.description);
        description = description ? description : '--';
        return <div className="ms-Grid-col ms-sm3"><Label aria-label={ariaLabel}>{this.props.settingModelDefinition.name} ({displayName} / {description})</Label></div>;
    }

    private readonly renderPropertySchema = (context: LocalizationContextInterface) => {
        const ariaLabel = context.t(ResourceKeys.deviceProperties.columns.schema);
        const { settingModelDefinition } = this.props;
        const schemaType = typeof settingModelDefinition.schema === 'string' ?
            settingModelDefinition.schema :
            settingModelDefinition.schema['@type'];
        return <div className="ms-Grid-col ms-sm2"><Label aria-label={ariaLabel}>{schemaType}</Label></div>;
    }

    private readonly renderPropertyUnit = (context: LocalizationContextInterface) => {
        const ariaLabel = context.t(ResourceKeys.deviceProperties.columns.unit);
        const unit = this.props.settingModelDefinition.unit;
        return <div className="ms-Grid-col ms-sm2"><Label aria-label={ariaLabel}>{unit ? unit : '--'}</Label></div>;
    }

    private readonly renderPropertyReportedValue = (context: LocalizationContextInterface) => {
        const { reportedTwin } = this.props;
        const ariaLabel = context.t(ResourceKeys.deviceSettings.columns.reportedValue);
        return (
            <div className="column-value-text ms-Grid-col ms-sm4" aria-label={ariaLabel}>
                <Stack horizontal={true} gap={'25px'}>
                    <Stack.Item align="start" className="reported-property">
                        {reportedTwin ?
                            (this.isSchemaSimpleType() ?
                                RenderSimplyTypeValue(
                                    reportedTwin.value,
                                    this.props.settingSchema,
                                    context.t(ResourceKeys.deviceSettings.columns.error)) :
                                <ActionButton
                                    className="column-value-button"
                                    ariaDescription={context.t(ResourceKeys.deviceSettings.command.openReportedValuePanel)}
                                    onClick={this.onViewReportedValue}
                                >
                                    {context.t(ResourceKeys.deviceSettings.command.openReportedValuePanel)}
                                </ActionButton>
                            ) : <Label>--</Label>
                        }
                    </Stack.Item>
                    {reportedTwin && reportedTwin.desiredState &&
                        <Stack.Item align="start">
                            {RenderDesiredState(reportedTwin.desiredState)}
                        </Stack.Item>
                    }
                </Stack>
            </div>
        );
    }

    private readonly renderCollapseButton = (context: LocalizationContextInterface) => {
        return (
        <div className="ms-Grid-col ms-sm1">
            <IconButton
                title={context.t(this.props.collapsed ? ResourceKeys.deviceSettings.command.expand : ResourceKeys.deviceSettings.command.collapse)}
                iconProps={{iconName: this.props.collapsed ? InterfaceDetailCard.OPEN : InterfaceDetailCard.CLOSE}}
            />
        </div>);
    }

    private readonly createUncollapsedCard = () => {
        return (
            <section className={this.props.collapsed ? 'item-detail' : 'item-detail item-detail-uncollapsed'}>
                {!this.props.collapsed && this.createForm()}
            </section>
        );
    }

    private readonly getSettingSchema = () => {
        const { settingModelDefinition } = this.props;
        return typeof settingModelDefinition.schema === 'string' ?
            settingModelDefinition.schema :
            settingModelDefinition.schema['@type'];
    }

    private readonly createReportedValuePanel = () => {
        const { reportedTwin, settingModelDefinition : modelDefinition, settingSchema : schema } = this.props;
        return (
            <>
            {
                this.state.showReportedValuePanel &&
                <div role="dialog">
                    <ComplexReportedFormPanel
                        schema={schema}
                        modelDefinition={modelDefinition}
                        showPanel={this.state.showReportedValuePanel}
                        formData={reportedTwin}
                        handleDismiss={this.handleDismissViewReportedPanel}
                    />
                </div>
            }
            </>
        );
    }

    private readonly onViewReportedValue = () => {
        this.setState({showReportedValuePanel: true});
        this.props.handleOverlayToggle();
        this.props.handleCollapseToggle();
    }

    private readonly handleDismissViewReportedPanel = () => {
        this.setState({showReportedValuePanel: false});
        this.props.handleOverlayToggle();
    }

    private readonly createForm = () => {
        return (
            <DataForm
                buttonText={ResourceKeys.deviceSettings.command.submit}
                formData={this.props.desiredTwin}
                settingSchema={this.props.settingSchema}
                handleSave={this.onSubmit}
                craftPayload={this.createSettingsPayload}
                interfaceName={this.props.interfaceName}
                schema={this.getSettingSchema()}
            />
        );
    }
    private readonly createSettingsPayload = (patchData: object) => {
        return generateInterfacePropertiesPayload(this.props.interfaceName, this.props.settingModelDefinition.name, patchData);
    }

    private readonly onSubmit = (twin: any) => () => { // tslint:disable-line:no-any
        this.props.patchDigitalTwinInterfaceProperties({
            digitalTwinId: this.props.deviceId,
            interfacesPatchData: twin,
            propertyKey: this.props.settingModelDefinition.name
        });
    }
}
