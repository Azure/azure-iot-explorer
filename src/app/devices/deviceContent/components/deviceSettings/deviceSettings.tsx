/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Shimmer } from 'office-ui-fabric-react/lib/Shimmer';
import { CommandBar } from 'office-ui-fabric-react/lib/CommandBar';
import { RouteComponentProps } from 'react-router-dom';
import DeviceSettingPerInterface from './deviceSettingsPerInterface';
import { TwinWithSchema } from './deviceSettingsPerInterfacePerSetting';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString, getInterfaceIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { PatchDigitalTwinInterfacePropertiesActionParameters } from '../../actions';
import InterfaceNotFoundMessageBoxContainer from '../shared/interfaceNotFoundMessageBarContainer';
import { REFRESH } from '../../../../constants/iconNames';

export interface DeviceSettingsProps extends DeviceInterfaceWithSchema{
    isLoading: boolean;
}

export interface DeviceInterfaceWithSchema {
    interfaceId: string;
    interfaceName: string;
    twinWithSchema: TwinWithSchema[];
}

export interface DeviceSettingDispatchProps {
    refresh: (deviceId: string, interfaceId: string) => void;
    setInterfaceId: (id: string) => void;
    patchDigitalTwinInterfaceProperties: (parameters: PatchDigitalTwinInterfacePropertiesActionParameters) => void;
}

export default class DeviceSettings
    extends React.Component<DeviceSettingsProps & DeviceSettingDispatchProps & RouteComponentProps, {}> {
    constructor(props: DeviceSettingsProps & DeviceSettingDispatchProps & RouteComponentProps) {
        super(props);
    }

    public render(): JSX.Element {
        if (this.props.isLoading) {
            return (
                <Shimmer/>
            );
        }

        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <>
                        <CommandBar
                                className="command"
                                items={[
                                    {
                                        ariaLabel: context.t(ResourceKeys.deviceSettings.command.refresh),
                                        iconProps: {iconName: REFRESH},
                                        key: REFRESH,
                                        name: context.t(ResourceKeys.deviceSettings.command.refresh),
                                        onClick: this.handleRefresh
                                    }
                                ]}
                        />
                        {this.renderProperties(context)}
                    </>
                )}
            </LocalizationContextConsumer>
        );
    }

    public componentDidMount() {
        this.props.setInterfaceId(getInterfaceIdFromQueryString(this.props));
    }

    private readonly renderProperties = (context: LocalizationContextInterface) => {
        return (
            <>
                <h3>{context.t(ResourceKeys.deviceSettings.headerText)}</h3>
                {this.props.twinWithSchema ?
                    <DeviceSettingPerInterface
                        {...this.props}
                        deviceId={getDeviceIdFromQueryString(this.props)}
                    /> :
                    <InterfaceNotFoundMessageBoxContainer/>
                }
            </>
        );
    }

    private readonly handleRefresh = () => {
        this.props.refresh(getDeviceIdFromQueryString(this.props), getInterfaceIdFromQueryString(this.props));
    }
}
