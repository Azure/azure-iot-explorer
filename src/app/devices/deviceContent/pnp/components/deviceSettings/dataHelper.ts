/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { JsonSchemaAdaptor } from '../../../../../shared/utils/jsonSchemaAdaptor';
import { ModelDefinition } from '../../../../../api/models/modelDefinition';
import { TwinWithSchema } from './deviceSettingsPerInterfacePerSetting';

export interface DeviceInterfaceWithSchema {
    twinWithSchema: TwinWithSchema[];
}

export const generateTwinSchemaAndInterfaceTuple = (model: ModelDefinition, digitalTwin: object, componentName: string): DeviceInterfaceWithSchema => {
    if (!model) {
        return { twinWithSchema: [] };
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(model);
    const writableProperties = jsonSchemaAdaptor.getWritableProperties();
    const digitalTwinForSpecificComponent = digitalTwin && (digitalTwin as any)[componentName]; // tslint:disable-line: no-any

    const settings = writableProperties
        .map(setting => {
            return {
                isComponentContainedInDigitalTwin: !!digitalTwinForSpecificComponent,
                metadata: digitalTwinForSpecificComponent && digitalTwinForSpecificComponent.$metaData && digitalTwinForSpecificComponent.$metadata[setting.name],
                reportedTwin: digitalTwinForSpecificComponent && digitalTwinForSpecificComponent[setting.name],
                settingModelDefinition: setting,
                settingSchema: jsonSchemaAdaptor.parseInterfacePropertyToJsonSchema(setting)
            };
        });

    return {
        twinWithSchema: settings,
    };
};
