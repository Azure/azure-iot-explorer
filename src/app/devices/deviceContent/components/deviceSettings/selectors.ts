/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from './../../../../api/models/modelDefinition';
import { PropertyContent, ContentType } from '../../../../api/models/modelDefinition';
import { StateInterface } from '../../../../shared/redux/state';
import { parseInterfacePropertyToJsonSchema } from '../../../../shared/utils/jsonSchemaAdaptor';
import { getModelDefinitionSelector, getComponentNameSelector } from '../../selectors';
import { DeviceInterfaceWithSchema } from './deviceSettings';
import { generateDigitalTwinForSpecificProperty } from './../deviceProperties/selectors';

export const getDeviceSettingTupleSelector = (state: StateInterface) => {
    const modelDefinition = getModelDefinitionSelector(state);
    return modelDefinition && generateTwinSchemaAndInterfaceTuple(state, modelDefinition);
};

const generateTwinSchemaAndInterfaceTuple = (state: StateInterface, model: ModelDefinition): DeviceInterfaceWithSchema => {
    const writableProperties = model && model.contents && model.contents.filter((item: PropertyContent) => filterProperties(item)) as PropertyContent[];

    const settings = writableProperties && writableProperties
        .map(setting => {
            const property = generateDigitalTwinForSpecificProperty(state, setting);
            return {
                desiredTwin: property && property.desired && property.desired.value,
                reportedTwin: property && property.reported,
                settingModelDefinition: setting,
                settingSchema: parseInterfacePropertyToJsonSchema(setting)
            };
        });

    return {
        componentName: getComponentNameSelector(state),
        interfaceId: model['@id'],
        twinWithSchema: settings,
    };
};

const filterProperties = (content: PropertyContent) => {
    if (typeof content['@type'] === 'string') {
        return content['@type'].toLowerCase() === ContentType.Property && content.writable;
    }
    else {
        return content['@type'].some((entry: string) => entry.toLowerCase() === ContentType.Property) && content.writable;
    }
};
