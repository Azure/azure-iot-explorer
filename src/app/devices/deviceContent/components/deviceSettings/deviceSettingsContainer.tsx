/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { compose, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { AnyAction } from 'typescript-fsa';
import { StateType } from '../../../../shared/redux/state';
import DeviceSettings, { DeviceSettingDispatchProps, DeviceSettingsProps } from './deviceSettings';
import { setInterfaceIdAction, patchDigitalTwinInterfacePropertiesAction, PatchDigitalTwinInterfacePropertiesActionParameters, getDigitalTwinInterfacePropertiesAction, getModelDefinitionAction } from '../../actions';
import { getDeviceSettingTupleSelector } from './selectors';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { getModelDefinitionSyncStatusSelector, getDigitalTwinInterfacePropertiesStateSelector } from '../../selectors';

const mapStateToProps = (state: StateType): DeviceSettingsProps => {
    return {
        isLoading: getDigitalTwinInterfacePropertiesStateSelector(state) === SynchronizationStatus.working ||
            getModelDefinitionSyncStatusSelector(state) === SynchronizationStatus.working,
        ...getDeviceSettingTupleSelector(state),
    };
};

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>): DeviceSettingDispatchProps => {
    return {
        patchDigitalTwinInterfaceProperties:
            (parameters: PatchDigitalTwinInterfacePropertiesActionParameters) => dispatch(patchDigitalTwinInterfacePropertiesAction.started(parameters)),
        refresh: (deviceId: string, interfaceId: string) => {
            dispatch(getDigitalTwinInterfacePropertiesAction.started(deviceId));
            dispatch(getModelDefinitionAction.started({digitalTwinId: deviceId, interfaceId}));
        },
        setInterfaceId: (id: string) => dispatch(setInterfaceIdAction(id))
    };
};

export default compose(withRouter, connect(mapStateToProps, mapDispatchToProps))(DeviceSettings);
