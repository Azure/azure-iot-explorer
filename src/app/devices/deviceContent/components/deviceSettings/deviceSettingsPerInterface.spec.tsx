/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import DeviceSettingsPerInterface, { DeviceSettingDataProps, DeviceSettingDispatchProps, DeviceSettingState } from './deviceSettingsPerInterface';
import { mountWithLocalization } from '../../../../shared/utils/testHelpers';
import { twinWithSchema } from './deviceSettings.spec';

describe('components/devices/deviceSettingsPerInterface', () => {

    const deviceSettingsProps: DeviceSettingDataProps = {
        deviceId: 'testDevice',
        interfaceId: 'urn:contoso:com:EnvironmentalSensor:1',
        interfaceName: 'environmentalSensor',
        twinWithSchema: [twinWithSchema]
    };

    const deviceSettingsDispatchProps: DeviceSettingDispatchProps = {
        patchDigitalTwinInterfaceProperties: jest.fn()
    };

    const getComponent = (overrides = {}) => {
        const props = {
            ...deviceSettingsProps,
            ...deviceSettingsDispatchProps,
            ...overrides
        };

        return mountWithLocalization(
            <DeviceSettingsPerInterface {...props} />
        );
    };

    it('matches snapshot', () => {
        expect(getComponent()).toMatchSnapshot();
    });

    it('toggles collapsed', () => {
        const wrapper = getComponent();
        expect((wrapper.state() as DeviceSettingState).allCollapsed).toBeFalsy();
        const button = wrapper.find(IconButton).at(1);
        button.simulate('click');
        wrapper.update();
        expect((wrapper.state() as DeviceSettingState).allCollapsed).toBeTruthy();
    });

    it('executes handle toggle from child', () => {
        const wrapper = getComponent();
        let collapsed = (wrapper.state() as DeviceSettingState).collapseMap.get(0);
        expect(collapsed).toBeFalsy();
        // tslint:disable-next-line:no-magic-numbers
        const button = wrapper.find(IconButton).at(2);
        button.simulate('click');
        wrapper.update();
        collapsed = (wrapper.state() as DeviceSettingState).collapseMap.get(0);
        expect(collapsed).toBeTruthy();
    });
});
