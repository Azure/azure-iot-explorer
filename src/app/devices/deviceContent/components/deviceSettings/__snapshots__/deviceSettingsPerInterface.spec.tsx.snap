// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/deviceSettings matches snapshot while interface cannot be found 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    farItems={
      Array [
        Object {
          "ariaLabel": "deviceSettings.command.close",
          "iconProps": Object {
            "iconName": "NavigateBack",
          },
          "key": "NavigateBack",
          "name": "deviceSettings.command.close",
          "onClick": [Function],
        },
      ]
    }
    items={
      Array [
        Object {
          "ariaLabel": "deviceSettings.command.refresh",
          "iconProps": Object {
            "iconName": "Refresh",
          },
          "key": "Refresh",
          "name": "deviceSettings.command.refresh",
          "onClick": [Function],
        },
      ]
    }
  />
  <Route
    component={[Function]}
  />
  <Connect(InterfaceNotFoundMessageBox) />
</Fragment>
`;

exports[`components/devices/deviceSettings matches snapshot with one twinWithSchema 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    farItems={
      Array [
        Object {
          "ariaLabel": "deviceSettings.command.close",
          "iconProps": Object {
            "iconName": "NavigateBack",
          },
          "key": "NavigateBack",
          "name": "deviceSettings.command.close",
          "onClick": [Function],
        },
      ]
    }
    items={
      Array [
        Object {
          "ariaLabel": "deviceSettings.command.refresh",
          "iconProps": Object {
            "iconName": "Refresh",
          },
          "key": "Refresh",
          "name": "deviceSettings.command.refresh",
          "onClick": [Function],
        },
      ]
    }
  />
  <Route
    component={[Function]}
  />
  <DeviceSettingsPerInterface
    componentName="environmentalSensor"
    deviceId={null}
    history={
      Object {
        "location": Object {
          "pathname": "/#/devices/deviceDetail/ioTPlugAndPlay/ioTPlugAndPlayDetail/settings/?id=device1&componentName=foo&interfaceId=urn:contoso:com:EnvironmentalSensor:1",
        },
      }
    }
    interfaceId="urn:contoso:com:EnvironmentalSensor:1"
    isLoading={false}
    location={
      Object {
        "pathname": "/#/devices/deviceDetail/ioTPlugAndPlay/ioTPlugAndPlayDetail/settings/?id=device1&componentName=foo&interfaceId=urn:contoso:com:EnvironmentalSensor:1",
      }
    }
    match={Object {}}
    patchDigitalTwinInterfaceProperties={[MockFunction]}
    refresh={[MockFunction]}
    setComponentName={
      [MockFunction] {
        "calls": Array [
          Array [
            null,
          ],
          Array [
            null,
          ],
          Array [
            null,
          ],
          Array [
            null,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    twinWithSchema={
      Array [
        Object {
          "desiredTwin": 123,
          "reportedTwin": null,
          "settingModelDefinition": Object {
            "@type": "Property",
            "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
            "displayName": "Brightness Level",
            "name": "brightness",
            "schema": "long",
            "writable": true,
          },
          "settingSchema": Object {
            "description": "Brightness Level / The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
            "title": "brightness",
            "type": "number",
          },
        },
      ]
    }
  />
</Fragment>
`;

exports[`components/devices/deviceSettingsPerInterface matches snapshot 1`] = `
<div
  className="pnp-detail-list scrollable-lg"
>
  <div
    className="list-header flex-grid-row"
  >
    <span
      className="col-sm3"
    >
      deviceSettings.columns.name
    </span>
    <span
      className="col-sm2"
    >
      deviceSettings.columns.schema
    </span>
    <span
      className="col-sm2"
    >
      deviceSettings.columns.unit
    </span>
    <span
      className="col-sm4 reported-value"
    >
      deviceSettings.columns.reportedValue
      <StyledTooltipHostBase
        calloutProps={
          Object {
            "gapSpace": 0,
          }
        }
        content="deviceSettings.columns.reportedValueTooltip"
        id="tooltipHost23"
        styles={
          Object {
            "root": Object {
              "display": "inline-flex",
            },
          }
        }
      >
        <CustomizedIconButton
          aria-labelledby="tooltipHost23"
          iconProps={
            Object {
              "iconName": "Info",
            }
          }
        />
      </StyledTooltipHostBase>
    </span>
    <div
      className="col-sm1 collapse-button"
    >
      <CustomizedIconButton
        ariaLabel="deviceSettings.command.collapseAll"
        iconProps={
          Object {
            "iconName": "ChevronUp",
          }
        }
        onClick={[Function]}
        title="deviceSettings.command.collapseAll"
      />
    </div>
  </div>
  <section
    className="list-content"
    role="list"
  >
    <DeviceSettingsPerInterfacePerSetting
      collapsed={false}
      componentName="environmentalSensor"
      desiredTwin={123}
      deviceId="testDevice"
      handleCollapseToggle={[Function]}
      handleOverlayToggle={[Function]}
      interfaceId="urn:contoso:com:EnvironmentalSensor:1"
      key="0"
      patchDigitalTwinInterfaceProperties={[MockFunction]}
      reportedTwin={null}
      settingModelDefinition={
        Object {
          "@type": "Property",
          "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
          "displayName": "Brightness Level",
          "name": "brightness",
          "schema": "long",
          "writable": true,
        }
      }
      settingSchema={
        Object {
          "description": "Brightness Level / The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
          "title": "brightness",
          "type": "number",
        }
      }
      twinWithSchema={
        Array [
          Object {
            "desiredTwin": 123,
            "reportedTwin": null,
            "settingModelDefinition": Object {
              "@type": "Property",
              "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
              "displayName": "Brightness Level",
              "name": "brightness",
              "schema": "long",
              "writable": true,
            },
            "settingSchema": Object {
              "description": "Brightness Level / The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
              "title": "brightness",
              "type": "number",
            },
          },
        ]
      }
    />
  </section>
</div>
`;
